emacs: enable xwidgets support on Darwin

tani opened the pull request:
- Remove Darwin exclusion from withXwidgets condition
- Use apple-sdk webkit instead of webkitgtk_4_0 on Darwin
- Add comment explaining Apple SDK webkit compatibility


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tani committed:
emacs: enable xwidgets support on Darwin

- Remove Darwin exclusion from withXwidgets condition
- Use apple-sdk webkit instead of webkitgtk_4_0 on Darwin
- Add comment explaining Apple SDK webkit compatibility

jian-lin reviewed (commented):
LGTM.  I would like someone familiar with darwin to have another look.

tani committed:
emacs: remove apple-sdk for buildInputs

emilazy commented:
While we’re getting people to test Emacs on macOS stuff… https://github.com/NixOS/nixpkgs/pull/435148 :)

amusingimpala75 commented:
You need to update the
```nix
assert withXwidgets -> !noGui && (withGTK3 || withPgtk);
```
on line 132 otherwise the assert still fails

amusingimpala75 reviewed (changes requested):
Also I think the default `withXwidgets` is a bit broken?

(Sorry if my previous comment is duplicated, I just wanted to include it in the review proper

tani committed:
emacs: enable xwidgets on Cocoa and MacPorts

Update the `--with-xwidgets` flag to also be
enabled when building with Cocoa (`withNS`) or the
MacPorts variant. The assertion is adjusted to
reflect this new condition, and the WebKitGTK
dependency is now limited to Linux only.

amusingimpala75 commented:
LGTM, it works if I build it with `nix run` and if I add it to my `home-manager` config, but through `emacs-overlay` they force `xwidgets` disabled by default which should probably change for `darwin`, but that's not an issue for this PR as once I `override` the derivation it works as expected.

panchoh reviewed (approved):
I can’t test the changes on macOS, but the logic seems solid to me.
Thanks for this fix, and thanks to the reviewers for keeping it accurate!

kfiz commented on pkgs/applications/editors/emacs/make-emacs.nix:
```diff
@@ -96,10 +96,11 @@
-    && (lib.versionOlder version "30"), # XXX: upstream bug 66068 precludes newer versions of webkit2gtk (https://lists.gnu.org/archive/html/bug-gnu-emacs/2024-09/msg00695.html)
+    !noGui
+    && (withGTK3 || withPgtk || withNS || variant == "macport")
+    && (stdenv.hostPlatform.isDarwin || lib.versionOlder version "30"),
```
@alanpearce Why are we switching this on by default on darwin? If we just want to enable it for the gui versions then the previous condition on line 100 should be ok.

kfiz replied:
@tani thanks for clarifying. Looks weird though, but I wouldn't have an alternative to achieve the same.

tani committed:
emacs: remove !noGui condition for withXwidgets

AndersonTorres commented:
Yes. The idea is, precisely, build Emacs with no graphical dependencies besides good old terminal.
This `noGui` flag is a generalization of `emacs-nox` flavor, given that Emacs supports other window toolkits (MS Windows, Mac and even BeOS), and arguably no one expects to, say, run emacs-nox in Haiku and ends with a graphical interface.

tani committed:
Update pkgs/applications/editors/emacs/make-emacs.nix

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

AndersonTorres commented on pkgs/applications/editors/emacs/make-emacs.nix:
```diff
@@ -129,7 +129,8 @@ assert withGpm -> stdenv.hostPlatform.isLinux;
 assert withNS -> stdenv.hostPlatform.isDarwin && !(withX || variant == "macport");
 assert withPgtk -> withGTK3 && !withX;
-assert withXwidgets -> !noGui && (withGTK3 || withPgtk);
+assert withXwidgets -> !noGui && (withGTK3 || withPgtk || withNS || variant == "macport");
```
I did not pay too much attention because it touches the Darwin ecosystem.

Indeed, I see no reason to remove `!noGui` as done above. Per consistency we should keep it in both.

tani committed:
Revert "emacs: remove !noGui condition for withXwidgets"

This reverts commit b20c6205719cf3d608947f6d15271641be559988.

tani commented:
Thank you for the feedback. I believe I have now addressed all the suggestions raised in the discussion. I'm hoping we can move forward with merging this. Please let me know if there's anything else needed from my side.

AndersonTorres commented:
Classic rebase + force-push. I vote for a single commit blob.

A good commit message, summarizing changes like a changelog, should be enough.

jian-lin reviewed (approved):
diff LGTM.  I did not test it since I am not a darwin user.
